src/coprocessor/dag/rpn_expr/impl_arithmetic.rs

```diff
@@ -0,0 +1,216 @@
+pub struct UintUintPlus;
+
+impl ArithmeticOp for UintUintPlus {
+    type T = Int;
+
+    fn calc(lhs: &Int, rhs: &Int) -> Result<Option<Int>> {
+        (*lhs as u64)
+            .checked_add(*rhs as u64)
+            .ok_or_else(|| {
+                Error::overflow("BIGINT UNSIGNED", &format!("({} + {})", lhs, rhs)).into()
+            })
+            .map(|v| Some(v as i64))
+    }
+}
+
+#[derive(Debug)]
+pub struct RealPlus;
+
+impl ArithmeticOp for RealPlus {
+    type T = Real;
+
+    fn calc(lhs: &Real, rhs: &Real) -> Result<Option<Real>> {
+        let res = lhs + rhs;
+        if res.is_infinite() {
+            Err(Error::overflow("DOUBLE", &format!("({} + {})", lhs, rhs)))?;
+        }
+        Ok(Some(res))
+    }
+}
+
+#[derive(Debug)]
+pub struct DecimalPlus;
+
+impl ArithmeticOp for DecimalPlus {
+    type T = Decimal;
+
+    fn calc(lhs: &Decimal, rhs: &Decimal) -> Result<Option<Decimal>> {
+        let res: codec::Result<Decimal> = (lhs + rhs).into();
+        Ok(Some(res?))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::coprocessor::dag::rpn_expr::types::test_util::RpnFnScalarEvaluator;
+    use cop_datatype::builder::FieldTypeBuilder;
+    use cop_datatype::{FieldTypeFlag, FieldTypeTp};
+    use std::i64;
+    use tipb::expression::ScalarFuncSig::*;
+
+    #[test]
+    fn test_arithmetic_int() {
+        let test_cases = vec![
+            (PlusInt, None, false, Some(1), false, None),
+            (PlusInt, Some(1), false, None, false, None),
+            (PlusInt, Some(17), false, Some(25), false, Some(42)),
+            (
+                PlusInt,
+                Some(i64::MIN),
+                false,
+                Some((i64::MAX as u64 + 1) as i64),
+                true,
+                Some(0),
+            ),
+        ];
+        for (sig, lhs, lhs_is_unsigned, rhs, rhs_is_unsigned, expected) in test_cases {
+            let lhs_field_type = FieldTypeBuilder::new()
+                .tp(FieldTypeTp::LongLong)
+                .flag(if lhs_is_unsigned {
+                    FieldTypeFlag::UNSIGNED
+                } else {
+                    FieldTypeFlag::empty()
+                })
+                .build();
+            let rhs_field_type = FieldTypeBuilder::new()
+                .tp(FieldTypeTp::LongLong)
+                .flag(if rhs_is_unsigned {
+                    FieldTypeFlag::UNSIGNED
+                } else {
+                    FieldTypeFlag::empty()
+                })
+                .build();
+            let output = RpnFnScalarEvaluator::new()
+                .push_param_with_field_type(lhs, lhs_field_type)
+                .push_param_with_field_type(rhs, rhs_field_type)
+                .evaluate::<Int>(sig)
+                .unwrap();
+            assert_eq!(output, expected, "{:?}, {:?}", output, expected);
+        }
+    }
+
+    #[test]
+    fn test_arithmetic_real() {
+        let test_cases = vec![(PlusReal, Some(1.01001), Some(-0.01), Some(1.00001))];
```

Comment:
I think we also need to test the case which will overflow.

Reply:
Real overflow case covered.